File: Javascript/DP/E-121. Best Time to Buy and Sell Stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let maxProfit = 0;
  let currMinPrice = Infinity;
  for (const price of prices) {
    currMinPrice = Math.min(currMinPrice, price);
    currProfit = price - currMinPrice;
    maxProfit = Math.max(maxProfit, currProfit);
  }

  return maxProfit;
};

const testData = [
  [7, 1, 5, 3, 6, 4], //5
  [7, 6, 4, 3, 1], //0
];

for (const data of testData) {
  console.log(maxProfit(data));
}

File: Python/DP/E-121. Best Time to Buy and Sell Stock.py
import sys


class Solution:
    @classmethod
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        max_profit = 0
        curr_min_price = sys.maxsize
        for price in prices:
            curr_min_price = min(price, curr_min_price)
            curr_profit = price - curr_min_price
            max_profit = max(max_profit, curr_profit)

        return max_profit


testData = [
    [7, 1, 5, 3, 6, 4],  # 5
    [7, 6, 4, 3, 1],  # 0
]

for data in testData:
    print(Solution.maxProfit(data))
